Hermetic build documentation (#146)
redhat-appstudio/docs.appstudio.io: docs/modules/ROOT/pages/how-to-guides/proc_prefetching-dependencies-to-support-hermetic-build.adoc

:_content-type: PROCEDURE

[id="prefetching-dependencies-to-support-hermetic-build_{context}"]
= Prefetching the package manager dependencies for the Hermetic build

> {ProductName} uses Cachi2 to prefetch the package manager dependencies for the <<supported-languages>>. Cachi2 is a wrapper around the package managers that makes it easy to prefetch and cache dependencies. It also helps generate a manifest of all dependencies included in your builds, ensuring transparency and maintainability.

Reviewer: Don't use these words necessarily, but I feel like we need some additional introduction to Cachi2 and prefetch from the previous procedure of enabling hermetic builds which may fail.
```suggestion
Since hermetic builds are enabled in {ProductName} by restricting network access to the build, many builds using common package managers will fail as they are unable to call out to the repositories to pull in dependencies. In order to assist with building these artifacts, {ProductName} uses Cachi2 to prefetch the package manager dependencies for the <<supported-languages>>. Cachi2 is a wrapper around the package managers that makes it easy to prefetch and cache dependencies. It also helps generate a manifest of all dependencies included in your builds, ensuring transparency and maintainability.
```

> {ProductName} uses Cachi2 to prefetch the package manager dependencies for the <<supported-languages>>. Cachi2 is a wrapper around the package managers that makes it easy to prefetch and cache dependencies. It also helps generate a manifest of all dependencies included in your builds, ensuring transparency and maintainability.

Reviewer 2: It's not doing any caching yet, that functionality will be added later with additional features. cc @nirzari

[#supported-languages]
.Supported languages

[cols="1,1"]
|===
|Language
|Package Manager

|Go
|`gomod`

|Python
|`pip`
|===

== Enabling prefetch builds for `gomod`

.Prerequisites
* You have an upgraded build pipeline.
* You have a `go.mod` file in your repository that lists all the the dependencies.

.Procedures
For each component that you want to build hermetically:

. Navigate to your component's source code.
. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.

+
[source,yaml]
----

> param:

Reviewer: ```suggestion
params:
```

    -  name: prefetch-input
       value: gomod <1>
----
<1> If the package manager lock file is located in the root directory of your project, enter `gomod`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in JSON format. For example,

> {"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}

Reviewer: You can resolve the `{<package-manager>}` here and more accurately specify the path to the go.mod file instead of the generic lockfile.

Author: Thoughts on this @arewm : Specify the package manager lock file relative to the package path. If the package manager lock file is located in the root of your repository, enter `gomod`. Otherwise, provide the path to the lock file in the JSON object format. For example, `{"type": "gomod", "path": "subpath/to/the/other/module"}`. Additionally, if you have multiple lock files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "pip"}, {"path": "subpath/to/other/module", "type": "pip"}]`. Here, `.` indicates that the package manager lock file is located in the root of your repository.

Reviewer: Maybe others won't agree with me, but I think the "simplicity" of just specifying the package manager does not have enough benefit to complicate matters when describing both methods. Instead, we should just ignore the fact that the "default" case is there (users can read the docs and use it if they want) or they can specify the `type` and `path`.

If you want to include both, just introduce it "advanced" way as if it were the only way and include a `NOTE` for the special case of go.mod being in the root.

You can even add an additional note about multiple package managers/paths if you want.

Author: @chmeliik @brunoapimentel Please let us know your thoughts on @arewm suggestion. 
As for me, if what we have is accurate, has the correct flow, and conveys what we want to convey, I don't think we should change it. Having said that, I'm open to suggestions and updates.

Reviewer 3: I tend to agree with @arewm here. While the `gomod` shorthand is nice when using `cachi2` as a CLI tool, it may only bring confusion for the RHTAP yaml.

It may be better to document only the JSON object and JSON array options

include::../partials/con_hermetic_verification.adoc[]

== Enabling prefetch builds for `pip`

> Cachi2 supports pip by parsing of `pip` requirements files, including but not limited to *requirements.txt* files placed in the root of your repository. By generically parsing `pip` requirements files, it downloads the specified dependencies.

Reviewer: ```suggestion
Cachi2 supports pip by parsing of `pip` requirements files, including but not limited to *requirements.txt* files placed in the root of your repository. By generically parsing `pip` requirements files, Cachi2 downloads the specified dependencies.
```

Reviewer: unresolving until the change is pushed.

Reviewer: Have you just made these changes locally? Can you unresolve the suggestion or push the fix?

> IMPORTANT: The requirements file can have a different name, as you can use multiple files to provide the dependencies. These requirements files function as lockfiles, encompassing all the transitive dependencies. It's crucial to actively pin each transitive dependency listed in the requirements file to a specific version.

Reviewer: ```suggestion
IMPORTANT: The requirements file can have a different name, as you can use multiple files to provide the dependencies. These requirements files function as lockfiles, encompassing all the transitive dependencies. It's crucial to actively pin each transitive dependency listed in the requirements file to a specific version and hash.
```

Reviewer: Have you just made these changes locally? Can you unresolve the suggestion or push the fix?

> IMPORTANT: The requirements file can have a different name, as you can use multiple files to provide the dependencies. These requirements files function as lockfiles, encompassing all the transitive dependencies. It's crucial to actively pin each transitive dependency listed in the requirements file to a specific version.

Reviewer 4: `crucial` is a strong word choice, but we don't actually state _why_ it's crucial or what the consequence is if users _don't_ actively pin each transitive dependency. Therefore, I suggest two things:

1. soften the word choice 
2. state why it's crucial. 

These options would look something like:

1. You must actively pin each transitive dependency listed in the requirements file to a specific version.
2. It's crucial that you actively pin each transitive dependency listed in the requirements file to a specific version. You must actively pin each transitive dependency because [state why]

> s, which include all the transitive dependencies. Each transitive dependency in the *requirements.txt* file must be pinned to a specific version.

Reviewer: I think this line was mistakenly included?

> s, which include all the transitive dependencies. Each transitive dependency in the *requirements.txt* file must be pinned to a specific version.

Reviewer 2: Looks like a typo or sentence fragment on this line.

Reviewer: Yeah, my similar comment was resolved, so I am _assuming_ that this is fixed locally and not pushed.

> s, which include all the transitive dependencies. Each transitive dependency in the *requirements.txt* file must be pinned to a specific version.

Reviewer 4: <img width="815" alt="Screenshot 2023-06-16 at 10 01 57 AM" src="https://github.com/redhat-appstudio/docs.appstudio.io/assets/70717303/12874f05-9dac-4844-8421-553091b98d6a">

See screenshot. Formatting is off here.

.Prerequisites
* You have an upgraded build pipeline.
* You have a <filename> file in your repository that lists all the the dependencies.

.Procedures
For each component that you want to build hermetically:

. Navigate to your component’s source code.
. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.

+
[source,yaml]
----

> param:

Reviewer 2: ```suggestion
params:
```
In my pipeline definitions, it's params, not param.

    -  name: prefetch-input

>        value: pip <1>

Reviewer 2: What should they be modifying in their definition? The params section? or the default section? I have three references per pipeline when searching for 'prefetch-input', one is empty, two are variable references.
```
    - default: ""
      description: Build dependencies to be prefetched by Cachi2
      name: prefetch-input
      type: string
```
```
       - name: prefetch-dependencies
      params:
      - name: input
        value: $(params.prefetch-input)
```
        
```
        - name: PREFETCH_INPUT
        value: $(params.prefetch-input)
```

Reviewer 3: They should be adding it to [.spec.params](https://github.com/containerbuildsystem/cachi2/blob/9917e6f3169bf74e661da1501417b8048a162204/.tekton/push.yaml#L12) I would say

```yaml
spec:
  params:
    - ...
    - name: prefetch-input
      value: ...
```

Reviewer: It needs to be added so you will not be able to search it to find it. Therefore, it might be helpful to direct the user to exactly the place to make the addition.

Reviewer 3: Shall I file a story for build-service to add all available params to the generated PipelineRuns in the user's .tekton dir?

Reviewer 3: Hmm, since the entire Pipeline is inlined in the PipelineRun, I guess it doesn't really matter

Normally, the params in a Pipeline declare the available params while those in a PipelineRun set the values for that particular "run" of a Pipeline

```python
# the Pipeline
def my_pipeline(hermetic: bool = False):
   ...

# the PipelineRun
my_pipeline(hermetic=True)
```
But this would matter more if the pipeline was only referenced (as it is [in our cachi2 repo](https://github.com/containerbuildsystem/cachi2/blob/main/.tekton/pull-request.yaml) - we probably created the PipelineRun manually). Since the entire pipeline definition is there, users might as well change the defaults

----
<1> If the package manager lock file is located in the root directory of your project, enter `pip`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in the JSON format. For example,
{"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}

> . Create a file (for example, pyproject.toml) with metadata (name and version) and place it in the root of your repository.

Reviewer 3: If your project is a real python project, then a metadata file will already exist. It's good to explain where cachi2 will look for the project metadata, but the wording may use some changes

Note: cachi2 currently does _require_ a metadata file, but we want to change that as soon as our intern gets back :smile:

Author: @chmeliik Do you agree we can update line # 77 to:

Review the root of your repository for a metadata file, for example, *pyproject.toml* file, *setup.py*, or *setup.cfg*. If there is no metadata file, create one, as Cachi2 looks for the name and version of your project in the metadata files.

Author: BTW, I've checked in this change, let me know if we are good.

Reviewer 3: Not necessarily the root of the repository; keep in mind that the user can specify the `path` to the package (by default, the path is `.`, the repo root). The metadata file should be at this path.

Otherwise sounds good to me

> . Create a file (for example, pyproject.toml) with metadata (name and version) and place it in the root of your repository.

Reviewer 4: I think we're allowed to be directive in our steps to users. 

4. Create a prproject.toml file with metadata, including name and version. 
5. Place this file in the root of your repository: 
[sample metadata you provide]

+
[source,metadata]
----
[metadata]
name = "my_package"
version = "0.1.0"
----

+
NOTE: Instead of a *pyproject.toml* file, you can also create a *setup.py* or *setup.cfg* file. For information on the metadata of these files, see link:https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#project-metadata[Project Metadata].

> . To generate a fully resolved requirements.txt file, run

Reviewer 2: Does it matter what kind of operating system they're on, or pip version they have, to be able to run this command? Do they need to add the generated file to their source repo, if so, where? the root? Do they have to generate this file? There isn't much context here, and requirements.txt wasn't referenced earlier in this procedure (other than listing that it could be processed at the start). It probably makes more sense to developers though.

Reviewer 3: > Does it matter what kind of operating system they're on

It does. The environment where the user runs this should be as close as possible to the environment in the container build. Same OS, same python version - that should usually be enough.

> Does it matter pip version they have

pip-compile is a separate tools from pip, it comes from the [pip-tools](https://github.com/jazzband/pip-tools) package. The package does [runtime-depend on pip](https://github.com/jazzband/pip-tools/blob/6d06be8dbd88298a9e62df1415a8da28a14eadc0/pyproject.toml#L42), but `pip install pip-tools` should do the right thing.

Reviewer: I had a heck of a time with the CATALOG when updating the requirements from a mac -- I had to manually revert a deletion every time because _my pip-compile_ dropped a linux requirement which was needed for CI.

Author: @mattreid @chmeliik  Do you think adding a Note just after the procedure in line # 90 is good?

The updated procedures will look like this:

> ...
> . Generate a fully resolved *requirements.txt* file that contains all the transitive dependencies and pins them to a specific version and hash by using the following command:
> 
> +
> [source,command]
> ----
> $ pip-compile pyproject.toml --generate-hashes
> ----

> NOTE(BULLET2):  To successfully run the command mentioned above your environment must be as close as possible to the environment in the container build. That is, the environment should have the same operating system and the same Python $major.$minor version.

> Add the *requirements.txt* to the root of your component’s source code. 

> ....

Reviewer 2: That seems better to me. You could consider if it's worth calling out that they'd need pip-tools installed, but maybe that's obvious and pip would help them realize and install that if they didn't anyway.

Reviewer 2: @chmeliik If they don't have pyproject.toml, but have either setup.py or setup.cfg, can they use that to generate requirements.txt? or to do the generation, they need the toml file? From the description here, all 3 sound equivalent for cachi2.

Reviewer 3: pip-compile does support all of them and cachi2 only cares about the generated requirements.txt

* https://pip-tools.readthedocs.io/en/latest/#requirements-from-pyproject-toml
* https://pip-tools.readthedocs.io/en/latest/#requirements-from-setup-py-and-setup-cfg
* https://pip-tools.readthedocs.io/en/latest/#requirements-from-requirements-in

_N.B. cachi2 does also look at pyproject.toml, setup.py and setup.cfg, but only to figure out the name and version of your project. For the list of dependencies, cachi2 always uses requirements.txt._

{"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}

. Create a file (for example, pyproject.toml) with metadata (name and version) and place it in the root of your repository.

+
[source,metadata]
----
[metadata]
name = "my_package"
version = "0.1.0"
----

+
NOTE: Instead of a *pyproject.toml* file, you can also create a *setup.py* or *setup.cfg* file. For information on the metadata of these files, see link:https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#project-metadata[Project Metadata].

. Generate a fully resolved requirements.txt file by using the following command:

+
[source,command]
----
$ pip-compile pyproject.toml --generate-hashes
----

+
NOTE: The command mentioned assumes that you have defined project dependencies in *pyproject.toml*. However, if you have defined the project dependencies in either the *setup.py* or *setup.cfg*, ensure to update the command accordingly.

. In the root of your repository create a file, *requirements-build.in*.

> . Copy the build system requirements from the *pyproject.toml* file to the *requirements-build.in* file.

Reviewer 4: For accessibility and usability, I think we should remove the image. Images are also too difficult to maintain and upkeep.

Author: Thank you for the suggestions, Jake!
For this, I believe we can keep the image, since it provides more/real context to the procedures.

:_content-type: PROCEDURE

[id="prefetching-dependencies-to-support-hermetic-build_{context}"]
= Prefetching the package manager dependencies for the Hermetic build

{ProductName} uses Cachi2 to prefetch the package manager dependencies for the <<supported-languages>>. Cachi2 is a wrapper around the package managers that makes it easy to prefetch and cache dependencies. It also helps generate a manifest of all dependencies included in your builds, ensuring transparency and maintainability. 

To ensure proper usage of Cachi2, follow these steps for your project’s builds process:

> * *Define dependencies* - Explicitly declare dependencies in a lock file generated by the package manager. Cachi2 only fetches dependencies that are declared in this file.

Reviewer: ```suggestion
* *Define dependencies* - Explicitly declare dependencies in lock files used by the package managers. Cachi2 only fetches dependencies that are declared in these files.
```

* *Reproducible builds* - Cachi2 enforces exact version matching for dependencies, including transitive dependencies. This ensures controlled and reproducible build dependencies.

> * *Secure builds* - While package managers use lock files to resolve dependencies automatically, it is crucial to verify the checksums of all the dependencies to safeguard against supply chain attacks.

Reviewer: These are not presented as actionable steps, instead they describe some of the Cachi2 model. What are the steps users need to do for their build processes?

Author: I've removed the Cachi2 content and provided a link to the Cachi2 docs.

[#supported-languages]
.Supported languages

[cols="1,1"]
|===
|Language
|Package Manager

|Go
|`gomod`

|Python
|`pip`
|===

== Enabling prefetch builds for `gomod`

.Prerequisites
* You have an upgraded build pipeline.

> * You have dependencies outside your git repository.

Reviewer: This isn't strictly a pre-requisite. You can run this without any dependencies. Maybe a pre-requisite could be that you have a `go.mod` file in your repository defining go dependencies?

.Procedures
For each component that you want to build hermetically:

. Navigate to your component's source code.
. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.

+
[source,yaml]
----
param:
    -  name: prefetch-input

>        value: <package-manager-lock-file> <1>

Reviewer: We are specifically describing the behavior for gomod, so we don't have to generalize it to an arbitrary package manager.

Also, this description is contrasting to the `<1>` below as the value may not be a path to a lock file, instead, it may just be the value `gomod`.

----

> <1> If the package manager lock file is located in the root directory of your project, enter `gomod`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in JSON format. For example,

Reviewer: Another use case is if there are multiple go.mod files in the repository.

{"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}

include::../partials/con_hermetic_verification.adoc[]

== Enabling prefetch builds for `pip`

> Cachi2 supports pip by parsing *requirements.txt* files placed in the root of your repository and then downloading the specified dependencies. These files should be lockfiles, which include all the transitive dependencies. Each transitive dependency in the *requirements.txt* file must be pinned to a specific version.

Reviewer: It does not only parse `requirements.txt` files, instead it generically parses pip requirements files. The requirement is the internal format, not the file name as multiple files can be used to provide the dependencies.

.Prerequisites
* You have an upgraded build pipeline.

> * You have dependencies outside your git repository.

Reviewer: Again, not strictly a requirement -- instead, the requirement is to have a file declaring dependencies.

Author: Removed the bullet as I've covered generating requirement.txt file and project metadata file in the procedures.

.Procedures
For each component that you want to build hermetically:

. Navigate to your component’s source code.
. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.

+
[source,yaml]
----
param:
    -  name: prefetch-input

>        value: <package-manager-lock-file> <1>

Reviewer: Same comment as above about the conflict between description and help text as well as the need to generalize further.

----
<1> If the package manager lock file is located in the root directory of your project, enter `pip`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in the JSON format. For example,

> {"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}

Reviewer: Specify this to pip. Instead of lockfile, maybe mention the path to the requirements file. @brunoapimentel @chmeliik, is it the path to the directory or the path to the actual file?

Reviewer 3: The `path` would be a path to the top-level directory to a package - the directory which contains pyproject.toml or, for older setuptools-based projects, setup.py or setup.cfg

Cachi2 will automatically look for requirements.txt and requirements-build.txt in this directory, but the user can override this. The whole input might looks like this

```jsonc
{
  "type": "pip",
  // path to the package (relative to the --source directory)
  // defaults to "."
  "path": ".",
  // specify requirements files (relative to the package path)
  // defaults to ["requirements.txt"] or [] if the file does not exist
  "requirements_files": ["requirements.txt", "requirements-extra.txt"],
  // specify *build* requirements files
  // defaults to ["requirements-build.txt"] or [] if the file does not exist
  "requirements_build_files": ["requirements-build.txt"],
}
```
(copied from https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#specifying-packages-to-process)

Reviewer: Thanks, Adam.
Have you just made these changes locally? Can you unresolve the suggestion or push the fix?

. Create a file (for example, pyproject.toml) with metadata (name and version) and place it in the root of your repository.

+
[source,metadata]
----
[metadata]
name = "my_package"
version = "0.1.0"
----

+
NOTE: Instead of a *pyproject.toml* file, you can also create a *setup.py* or *setup.cfg* file. For information on the metadata of these files, see link:https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#project-metadata[Project Metadata].

> . To generate a fully resolved requirements.txt file, run

Reviewer: What does it mean to be fully resolved? Fully resolved transitive runtime dependencies and hash pinning?

Reviewer: Have you just made these changes locally? Can you unresolve the suggestion or push the fix?

+
[source,command]
----
$ pip-compile pyproject.toml --generate-hashes
----

+

> NOTE: The command mentioned assumes that you have defined project dependencies in *pyproject.toml*. However, if you have defined the project dependencies in either the *setup.py* or *setup.cfg*, ensure to update the command accordingly.

Reviewer: I have often seen `pip-compile` used with a `requirements.in` file. Is this not a suggested possible path?

Author: @arewm your thoughts on this:
The command mentioned assumes that you have defined project dependencies in *pyproject.toml*. However, if you have defined the project dependencies in either the *setup.py*,  *requirements.txt*, or `requirements.in` files, ensure to update the command accordingly.

Reviewer: I think something like that should be fine.

include::../partials/con_hermetic_verification.adoc[]

== Enabling prefetch builds for `pip` with source dependencies

> If your project depends on `pip` with source dependencies, you can prefetch builds to improve the performance of the build process. Prefetch builds allow Cachi2 to fetch the build dependencies and provide them to Pip for offline installation in a network-isolated environment.

Reviewer: Will users know what "source dependencies" are? Are you just trying to say build dependencies?

I don't think that this is a matter of improving performance of builds as much as it is making builds possible that would otherwise fail due to the lack of availability of build-time dependencies.

Author: Thank you!
@arewm So do you think we can update the title to: `Enabling prefetch builds for `pip` with build dependencies`
And the text to: `If your project depends on `pip` with build dependencies, you can....`

Reviewer 2: If they were also cached, that could help subsequent builds, but to echo Andrew here, I don't think pre-fetching is going to speed anything up.

Author: @arewm So, update the title to: `Enabling prefetch builds for package dependencies in pip`
And the text to: `If your project has package dependencies in pip, you can....`

Reviewer: I think that sounds reasonable, yes.

.Prerequisites

* You have an upgraded build pipeline.
* You have dependencies outside of your git repository.

> * Download the link:https://raw.githubusercontent.com/containerbuildsystem/cachito/master/bin/pip_find_builddeps.py[`pip_find_builddeps.py`] script directly from GitHub.

Reviewer: What does this script achieve? Shouldn't this live below in the procedures as it automates part of the work of identifying the build dependencies?

Author: Thank you!

@chmeliik @brunoapimentel 
1. It's listed as a prerequisite here, do you think we can move it in the procedures? https://github.com/containerbuildsystem/cachi2/blob/806f868947382aef0ddf4a14fd7f3996274bfcfc/docs/pip.md#requirements-buildtxt

2. What is the location we need to save this file after downloading it?

Reviewer 3: > It's listed as a prerequisite here, do you think we can move it in the procedures?

yes :slightly_smiling_face: 

Downloading the script is a prerequisite to generating requirements-build.txt, but the whole thing is part of the procedure that prepares a repo

> What is the location we need to save this file after downloading it?

* anywhere in $PATH and make it executable
* or anywhere at all and run it as `python path/to/pip_find_builddeps.py <args>`

Reviewer: I think it makes sense to download the script as part of the procedure as it is optional to the overall process and only required to some specific part of the process.

+
NOTE: This script has no runtime dependency other than `pip`.

.Procedures
For each component that you want to build hermetically:

. Navigate to your component's source code.
. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.
. *Set up dependency prefetching with Cachi2:* In both the `.yaml` files, add the following prefetch-input pipeline parameter.

+
[source,yaml]
----
param:
    -  name: prefetch-input
       value: <package-manager-lock-file> <1>
----
<1> If the package manager lock file is located in the root directory of your project, enter `pip`. Otherwise, if the package manager lock file is in a subdirectory, provide the path to the lock file in the JSON format. For example,

> {"type": "{<package-manager>}", "path": "path-to-the-directory-which-contains-the-lockfile}"}

Reviewer: We will be suggesting users to have a requirements.txt and requirements-build.txt file below. We don't need all of this placeholder information, instead we can show an example of a more complete example.

Author: @arewm Thoughts on this: Specify the requirements files relative to the package path. If the requirements file is located in the root of your repository, enter `pip`. Otherwise, provide the path to the requirements file in the JSON object format. For example, `{"type": "pip", "path": "subpath/to/the/other/module"}`. Additionally, if you have multiple requirements files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "pip"}, {"path": "subpath/to/other/module", "type": "pip"}]`. Here, `.` indicates that the requirements file is located in the root of your repository.

Reviewer: We touched on multiple lock files above. We can either touch on it again here or just mention it once. I think just having one path is fine for this example with a path like you suggested ... `subpath/to/the/requirements/directory`

. Create a file (for example, pyproject.toml) with metadata (name and version) and place it in the root of your repository.

+
[source,metadata]
----
[metadata]
name = "my_package"
version = "0.1.0"
----

+
NOTE: Instead of a *pyproject.toml* file, you can also create a *setup.py* or *setup.cfg* file. For information on the metadata of these files, see link:https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#project-metadata[Project Metadata].

. Generate a fully resolved requirements.txt file by using the following command:

+
[source,command]
----
$ pip-compile pyproject.toml --generate-hashes
----

+
NOTE: The command mentioned assumes that you have defined project dependencies in *pyproject.toml*. However, if you have defined the project dependencies in either the *setup.py* or *setup.cfg*, ensure to update the command accordingly.

. In the root of your repository create a file, *requirements-build.in*.

> . Copy the build system requirements from the *pyproject.toml* file to the *requirements-build.in* file.

Reviewer: Should users know how to achieve this?

Author: I'm not sure if I get this - 
We already have listed how to create a *pyproject.toml* file.
We have an example image as well specifying the build system requirements

Please let me know what you think?

Author: I've added this image as an example: https://github.com/redhat-appstudio/docs.appstudio.io/pull/146/files#diff-a2672955c8abed8444b9800d148794fb076f60b817f810dd6819fbc2a681de4d

Please let me know if we need to add more information on this.

Reviewer: I guess the image is fine. If there is any documentation (i.e. within a pip site or about a PEP specification) that indicates the build requirements, then you can link to that as well.

+
*For example:*
+
image::build-requirements.png[Build requirements, role="image"]

. Run the `pip_find_builddeps.py` script and `pip-compile` the outputs by using the following command:

+
[source,command]
----
$ pip_find_builddeps.py requirements.txt \
--append \
--only-write-on-update \
-o requirements-build.in
----

. Use the `pip-compile` command to convert the *requirements-build.in* file in to the *requirements-build.txt* file by using the following command:

+
[source,command]
----
$ pip-compile requirements-build.in --allow-unsafe --generate-hashes
----

. Add the *requirement-build.txt* file to your project. It does not require any changes to your build process. 

+

> NOTE: `pip` automatically installs the build dependencies when needed for explicit installation. The purpose of the *requirement-build.txt* file is to enable Cachi2 to fetch the build dependencies and provide them to `pip` for offline installation in a network-isolated environment.

Reviewer: I don't think that this is always true. Sometimes changes are needed to the build process so that certain build-time dependencies can be built first before they are needed.

Author: @arewm Your thoughts on this: 
NOTE: `pip` automatically installs the necessary build dependencies when explicitly installing them. However, it's important to note that in certain scenarios, modifications to the build process may be required to ensure that specific build-time dependencies are built before they are actually required. In such cases, you can use the `requirements.txt` file to specify the build dependencies that must be installed. The purpose of the *requirement-build.txt* file is to enable Cachi2 to fetch the build dependencies and provide them to `pip` for offline installation in a network-isolated environment.

Reviewer 3: I think it's true enough to leave out the extra complexity. Pip - at least up to date versions of pip - does always install build dependencies automatically.

There are of course situations, especially if setuptools and pip in your base image are very outdated, where you can run into baffling bugs in build dependency handling

Reviewer: ack, thanks. I am fine leaving out for now then. Can we represent this in https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#troubleshooting so that users may be able to discover it without polluting this documentation?

Reviewer 3: It's meant to be https://github.com/containerbuildsystem/cachi2/blob/main/docs/pip.md#miscellaneous-errors-while-building-from-source

Reviewer: Fair enough -- it isn't directly what I was thinking of for a solution suggestion, but we can always update that. I think that this thread can be resolved without further actions. Any objections?

Author: I am marking this thread as resolved without any changes to the content.

:_content-type: PROCEDURE
:myfunctiontwo: prefetch_hermetic_builds

[id="prefetching-dependencies-to-support-hermetic-build_{context}"]
= Prefetching the package manager dependencies for the Hermetic build

> In {ProductName}, hermetic builds are achieved by restrciting network access to the build. Consequently, builds using supported languages may fail if they cannot retrieve dependencies from respositories. To address this, {ProductName} uses Cachi2 to prefetch the package manager dependencies for the supported-languages. Cachi2 makes it easier for package managers to prefetch dependencies. It also helps generate a manifest of all dependencies included in your builds, ensuring transparency and maintainability.

Reviewer 5: Minor typofixes here:

```suggestion
In {ProductName}, hermetic builds are achieved by restricting network access to the build. Consequently, builds using supported languages may fail if they cannot retrieve dependencies from repositories. To address this, {ProductName} uses Cachi2 to prefetch the package manager dependencies for the supported-languages. Cachi2 makes it easier for package managers to prefetch dependencies. It also helps generate a manifest of all dependencies included in your builds, ensuring transparency and maintainability.
```

[#supported-languages]
.Supported languages

[cols="1,1"]
|===
|Language
|Package Manager

|Go
|`gomod`

|Python
|`pip`
|===

== Enabling prefetch builds for `gomod`

.Prerequisites
* You have an link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/how-to-guides/configuring-builds/proc_upgrade_build_pipeline/[upgraded build pipeline].
* You have a `go.mod` file in your repository that lists all the the dependencies.

.Procedures
To create a hermetic build for a component, complete the following steps:

. Go to your component's source code.
. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.

> . Configure the hermetic pipeline by adding the following hermetic pipeline parameters in both the *.yaml* files:

Reviewer 2: @arewm @chmeliik @gtrivedi88 Sorry, I don't think we brought this discussion back here, but from the discussion in a chat https://redhat-internal.slack.com/archives/C04GVLR0155/p1686920028643609 having them set spec.params is sufficient, but our default yaml we provide them includes in-lined values, which would be overridden if they add the same name to spec.params. If users don't understand that hierarchy, they might get confused and change more than necessary, or find and replace those existing values. 

You could consider a note in the procedures where they're adding these params that they can ignore the related pipelinespec default value.

+
[source,yaml]
----
spec:
    params:
        -   ...
        -   name: prefetch-input
            value: gomod <1>
----
<1> Specify the package manager lock file relative to the package path. If the package manager lock file is located in the root of your repository, enter `gomod`. Otherwise, provide the path to the lock file in the JSON object format. For example: `{"type": "gomod", "path": "subpath/to/the/other/module"}`. Additionally, if you have multiple lock files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "pip"}, {"path": "subpath/to/other/module", "type": "pip"}]`. Here, `.` indicates that the package manager lock file is located in the root of your repository.

> . Commit your changes to the repository of the component.

Reviewer 2: Same comment here.

. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.

. Configure the hermetic pipeline by adding the following hermetic pipeline parameters in both the *.yaml* files:

+
[source,yaml]
----
spec:
    params:
        -   ...
        -   name: prefetch-input
            value: pip <1>
----
<1> Specify the requirements files relative to the package path. If the requirements file is located in the root of your repository, enter `pip`. Otherwise, provide the path to the requirements file in the JSON object format. For example, `{"type": "pip", "path": "subpath/to/the/other/module"}`. Additionally, if you have multiple requirements files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "pip"}, {"path": "subpath/to/other/module", "type": "pip"}]`. Here, `.` indicates that the requirements file is located in the root of your repository.

> . Commit your changes to the repository of the component.

Reviewer 2: They'd need to create a PR, or have that commit merge to master for it to build.

. Open the *.tekton* directory and locate the *.yaml* files related to the *pull-request* and *push* processes.

. Configure the hermetic pipeline by adding the following hermetic pipeline parameters in both the *.yaml* files:

+
[source,yaml]
----
spec:
    params:
        -   ...
        -   name: prefetch-input
            value: pip <1>
----
<1> Specify the requirements files relative to the package path. If the requirements file is located in the root of your repository, enter `pip`. Otherwise, provide the path to the requirements file in the JSON object format. For example, `{"type": "pip", "path": "subpath/to/the/other/module"}`. Additionally, if you have multiple requirements files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "pip"}, {"path": "subpath/to/the/requirements/directory", "type": "pip"}]`. Here, `.` indicates that the requirements file is located in the root of your repository.

> . Commit your changes to the repository of the component.

Reviewer 2: Same comment here as above. A build will only happen if there's a PR to the branch that component was configured for in RHTAP, or a merge to that branch.

[#supported-languages]
.Supported languages

[cols="1,1"]
|===
|Language
|Package Manager

|Go
|`gomod`

|Python
|`pip`
|===

== Enabling prefetch builds for `gomod`

.Prerequisites
* You have an link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/how-to-guides/configuring-builds/proc_upgrade_build_pipeline/[upgraded build pipeline].

* You have a `go.mod` file in your repository that lists all the dependencies.

.Procedures
To create a hermetic build for a component, complete the following steps:

. Go to the `.tekton` directory in the repository of your component and locate the *.yaml* files related to the *pull-request* and *push* processes.

. Configure the hermetic pipeline by adding the following hermetic pipeline parameters in both the *.yaml* files:

+
[source,yaml]
----
spec:
    params:
        -   ...
        -   name: prefetch-input
            value: '{"type": "gomod", "path": "."}' <1>
----

> <1> The `prefetch-input` parameter specifies the package manager lock file relative to the package path. In the above example, the `.` indicates that the package manager lock file is located in the root of the repository. Additionally, if you have multiple lock files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "gomod"}, {"path": "subpath/to/the/other/repository", "type": "gomod"}]`.

Reviewer 3: > specifies the package manager lock file relative to the package path

The `path` is a path to a directory, not to the lockfile. I would say it is the path to the package (where "package" means the directory that contains the lockfile and other package metadata files)

+
NOTE: When adding the parameters mentioned above, you can safely ignore the default values for the link:https://github.com/burrsutter/partner-catalog-stage/blob/e2ebb05ba8b4e842010710898d555ed3ba687329/.tekton/partner-catalog-stage-wgxd-pull-request.yaml#L90[`pipelineSpec.params`] in the *.yaml* files.

. Create a pull request (PR) by committing your changes to the repository of the component.

. Review and merge the PR.

include::../partials/con_hermetic_verification.adoc[]

== Enabling prefetch builds for `pip`
Cachi2 supports pip by parsing of `pip` requirements files, including but not limited to, *requirements.txt* files placed in the root of your repository. By generically parsing `pip` requirements files, Cachi2 downloads the specified dependencies.

> IMPORTANT: The requirements file can have a different name, as you can use multiple files to provide the dependencies. These requirements files function as lock files, encompassing all the transitive dependencies. You must actively pin each transitive dependency listed in the requirements file to a specific version.

Reviewer 3: For requirements files with non-default names/paths, the user would have to configure them in the prefetch-input

```json
{
  "type": "pip",
  "path": ".",
  "requirements_files": ["requirements.txt", "requirements-extras.txt", "tests/requirements.txt"]
}
```
The defaults are `requirements.txt` + `requirements-build.txt`

+
[source,yaml]
----
spec:
    params:
        -   ...
        -   name: prefetch-input
            value: '{"type": "pip", "path": "."}' <1>
----
<1> The `prefetch-input` parameter specifies the package manager lock file relative to the package path. In the above example, the `.` indicates that the package manager lock file is located in the root of the repository. Additionally, if you have multiple lock files, you can provide the path to the lock file in the JSON array format. For example, `[{"path": ".", "type": "pip"}, {"path": "subpath/to/the/other/repository", "type": "pip"}]`.

+
NOTE: When adding the parameters mentioned above, you can safely ignore the default values for the link:https://github.com/burrsutter/partner-catalog-stage/blob/e2ebb05ba8b4e842010710898d555ed3ba687329/.tekton/partner-catalog-stage-wgxd-pull-request.yaml#L90[`pipelineSpec.params`] in the *.yaml* files.

. Create a PR by committing your changes to the repository of the component.

. Review and merge the PR.

include::../partials/con_hermetic_verification.adoc[]

== Enabling prefetch builds for package dependencies in `pip`

> If your project has package dependencies in `pip`, you can prefetch builds to improve the performance of the build process. Prefetch builds allow Cachi2 to fetch the build dependencies and provide them to Pip for offline installation in a network-isolated environment.

Reviewer 3: I don't understand why "builds for package dependencies" are a separate section. If your project doesn't have any dependencies, you don't need to prefetch anything

Reviewer 3: Oh, this is about build dependencies. Build dependencies are dependencies which are only required while building your project and can be thrown away afterwards. The user always needs them, it's not optional. It also is not related improving performance. In fact, building from source is much slower than using pre-built dependencies. But building from source is the only way at the moment.

.Prerequisites

* You have an link:https://redhat-appstudio.github.io/docs.appstudio.io/Documentation/main/how-to-guides/configuring-builds/proc_upgrade_build_pipeline/[upgraded build pipeline].

* You have an environment that closely matches the environment in the container build, meaning it has the same operating system and the same python $major.$minor version.

* You have installed the link:https://github.com/jazzband/pip-tools[pip-tools] package.

.Procedures
To create a hermetic build for a component, complete the following steps:

> . Download the link:https://raw.githubusercontent.com/containerbuildsystem/cachito/master/bin/pip_find_builddeps.py[`pip_find_builddeps.py`] script directly from GitHub.

Reviewer 3: For some reason, this link doesn't look like a link in my browser (Firefox)

![image](https://github.com/redhat-appstudio/docs.appstudio.io/assets/38193343/6ba2cfe5-8d1c-4a7f-b323-a5d484f5b92a)
